### DISCUSSION TITLE
Should tags trigger a release?

### DISCUSSION THREAD

**Discussion by ghost** [fluxcd/flux2#1747] [Q&A] [answered]
If pushing a tag triggers changes to my cluster, and I can just delete the tag, is this even GitOps?

Shouldn't changes only be associated with commits?

**Comment by kingdonb** [comment, marked as answer]
I think there's a kernel of wisdom in here, even if this is a drive-by question and you deleted your account... Flux is in many ways not strongly opinionated and it supports lots of modes of operating. Some of them are going to be slightly incongruous with each other, and some of them might not even honor the principles of GitOps. It is on the Flux user to be informed and educated to avoid these modes. (We have a more constrained project that is more opinionated, Weave GitOps, but that's a different topic for another day...)

If you push a tag and delete it, this is not something that Flux especially has the power to control. Flux can be configured to watch for new tags, or to watch for new commits that advance the HEAD of a branch. (Or, sync a specific commit hash, only.) These are policy decisions. A separate policy decision is whether to allow or reject changes that challenge the immutability of artifacts on your git repo.

For example, a sensible policy would reject any "force push" that affects the main branch. Another sensible policy would be to reject updates that would change a tag that has already been published.

There is a use case for deleting tags as you describe: when a release is published that turns out to be dangerous, may cause data loss, and to be sure that nobody uses that release again. But if the policy permits a git tag to be deleted, (1) you might write a new tag with the same name, which has nearly unlimited potential to cause confusion down the line (don't do this), and (2) this also does not effectively guarantee that users will not receive this release, unless any downstream binary releases that were generated are also removed.

Rather than deleting a tag in git, for reasons outlined above which explain why I think this is a bad idea; if you strongly don't want a release to be used anymore I would leave the git tag intact, delete the image tag, and push a new release with a higher version number. But hopefully you can see how much of this is a policy decision that must be enforced at the Git Provider side, on the Image Repo, or somewhere else that is outside of the scope of Flux's control.

Users can set up branch and tag protection rules on a Git Repository with most providers like GitHub and GitLab, this is recommended. Here is a nice document about branch protection rules and the best practices for them, from GitHub: https://github.community/t/best-practices-for-protected-branches/10204
